src/poprox_storage/repositories/data_stores/db.py

```diff
@@ -16,10 +18,67 @@
 logger.setLevel(logging.DEBUG)
 
 
+DbRepositories: type[BaseModel] = create_model(
+    "DbRepositories",
+    conn=(Connection, None),
+    __config__=ConfigDict(arbitrary_types_allowed=True),
+)
+
+
+def camel_to_snake(s):
+    return "".join(["_" + c.lower() if c.isupper() else c for c in s]).lstrip("_")
+
+
+def db_repositories(*args):
+    if len(args) == 1 and not isinstance(args[0], list):
+        names = [args[0]]
+    else:
+        names = args
+
+    def inner_decorator(handler):
+        @wraps(handler)
+        def wrapper(event, context):
+            type_hints = get_type_hints(DbRepositories)
+            with DB_ENGINE.connect() as conn:
+                # create Repositories object and pre-construct all requested repository objects
+                repos = DbRepositories(conn=conn)
+                for name in names:
+                    repo_type = type_hints[name]
+                    setattr(repos, name, repo_type(conn))
+                return handler(event, context, repos)
+
+        return wrapper
+
+    return inner_decorator
+
+
 class DatabaseRepository:
+    _repository_types = {}
+
     def __init__(self, connection: Connection):
         self.conn: Connection = connection
 
+    def __init_subclass__(cls, *args, **kwargs):
+        """
+        Gets called once for each loaded class that sub-classes DatabaseRepository
+        """
+        global DbRepositories
+
+        # Register the attribute name and corresponding class
+        concept_name = camel_to_snake(cls.__name__.removeprefix("Db").removesuffix("Repository"))
+        cls._repository_types[concept_name] = cls
+
+        # Convert the attribute names and classes to fields
+        fields = {name: Annotated[type_, Field(default=None)] for name, type_ in cls._repository_types.items()}
+
+        # Dynamically create a Pydantic model with those fields
+        DbRepositories = create_model(
+            "DbRepositories",
+            conn=(Connection, None),
+            **fields,
+            __config__=ConfigDict(arbitrary_types_allowed=True),
+        )
```

Comment:
Every time we find another sub-class we create a new model type. That's not ideal but we don't know that we're done on any particular invocation of this hook. It might be possible to defer this until the decorator gets called though.